Framework/API/src/WorkspaceGroup.cpp

```diff
@@ -278,21 +279,33 @@ void WorkspaceGroup::workspaceDeleteHandle(
 }
 
 /**
- * Callback when a after-replace notification is received
- * Replaces a member if it was replaced in the ADS.
- * @param notice :: A pointer to a workspace after-replace notificiation object
+ * Callback when a before-replace notification is received
+ * Replaces a member if it was replaced in the ADS and checks
+ * for duplicate members within the group
+ * @param notice :: A pointer to a workspace before-replace notification object
  */
-void WorkspaceGroup::workspaceReplaceHandle(
+void WorkspaceGroup::workspaceBeforeReplaceHandle(
     Mantid::API::WorkspaceBeforeReplaceNotification_ptr notice) {
   std::lock_guard<std::recursive_mutex> _lock(m_mutex);
 
-  const std::string replacedName = notice->objectName();
+  const auto oldObject = notice->oldObject();
+
   for (auto &workspace : m_workspaces) {
-    if ((*workspace).name() == replacedName) {
+    if (workspace == oldObject) {
       workspace = notice->newObject();
       break;
     }
   }
+
+  // Remove any duplicate pointers
```

Comment:
I think we're nearly there with this now but I think there is a bug here.

The docs for [std::unique](http://en.cppreference.com/w/cpp/algorithm/unique) states that it only removes *consecutive* duplicates so if I updated your unit test as below to include a 3rd member

```
 void testRenameWorkspaceDuplicates() {
    // Tests that setting a workspace to the name of an existing
    // workspace in a group is correctly handled by the workspace group

    constexpr int nHist = 20, nBins = 10;
    // Register the workspace in the data service
    MatrixWorkspace_sptr work_in1 =
        WorkspaceCreationHelper::Create2DWorkspace123(nHist, nBins);
    MatrixWorkspace_sptr work_in2 =
        WorkspaceCreationHelper::Create2DWorkspace154(nHist, nBins);
    MatrixWorkspace_sptr work_in3 =
        WorkspaceCreationHelper::Create2DWorkspace154(nHist, nBins);

    const std::string ws1Name = "test_ws1";
    const std::string ws2Name = "test_ws2";
    const std::string ws3Name = "test_ws3";

    WorkspaceGroup_sptr wsSptr = WorkspaceGroup_sptr(new WorkspaceGroup);
    if (wsSptr) {
      AnalysisDataService::Instance().add("test_group", wsSptr);
      AnalysisDataService::Instance().add(ws1Name, work_in1);
      wsSptr->add(ws1Name);
      AnalysisDataService::Instance().add(ws2Name, work_in2);
      wsSptr->add(ws2Name);
      AnalysisDataService::Instance().add(ws3Name, work_in3);
      wsSptr->add(ws3Name);
    }

    TS_ASSERT_EQUALS(wsSptr->getNumberOfEntries(), 3);

    // Rename workspace 3 to 1
    AnalysisDataService::Instance().rename(ws3Name, ws1Name);

    TS_ASSERT_EQUALS(wsSptr->getNumberOfEntries(), 2);
    TS_ASSERT_EQUALS(wsSptr->contains(work_in3), true);
    TS_ASSERT_EQUALS(wsSptr->contains(work_in2), true);
    TS_ASSERT_EQUALS(wsSptr->contains(work_in1), false);

  }
```
then this test fails. I would suggest updating the unit test to include a 3rd member like this. 

It looks like you'd need to sort the vector for `std::unique` to work as we would want here. I thinking sorting shouldn't be a problem here.

I think it is also more idiomatic to use `std::unique` coupled with `std::erase` rather than `resize` as in the link above.